### ISSUE TITLE
using axicli from within processing.org sketch

### ISSUE THREAD

**Issue by stephanschulz** [evil-mad/axidraw#90]
I am working on a processing app that can prepare SVGs and use axicli to get them plotted.

Here is my code on how to call UNIX commands from within processing:
https://discourse.processing.org/t/example-running-unix-command-from-processing-to-execute-python-scripts/24285

Basically:
` String commandToRun = "/usr/local/bin/python3 axicli.py /Users/stephanschulz/Documents/axidraw-master/cli/AxiDraw_trivial.svg";`

This worked well all day.
And I also think this worked well when I was at work today:
`String commandToRun = "/usr/local/bin/python3 axicli.py version";`
But now at home without the axidraw machine here I get the bellow error when trying to ask axicli for its version.
I don't think it has anything todo with the fact that the machine is not connected to USB, right?

```
Stephan-2:~ stephanschulz$ /usr/local/bin/python3 axicli version
/usr/local/bin/python3: can't open file 'axicli': [Errno 2] No such file or directory
Stephan-2:~ stephanschulz$ 
```

Usually from the terminal I only need to call `axicli version` but from within processing I needed to specific the python3 path. 
The above error is the same as I get when running it in processing.

What am I doing wrong?

**Comment by stephanschulz** [comment]
figured it out.

it is important for line to point to the folder that contains the axidraw_api files.

Here the line I had to change for it to really point to the api folder:
` File workingDir = new File("/Users/stephanschulz/Documents/Processing/shell/test_axicli3/axidraw_api");`

Here my processing function.
```
void runCommand() {

  String commandToRun = "/usr/local/bin/python3 axicli.py version";

  File workingDir = new File("/Users/stephanschulz/Documents/Processing/shell/test_axicli3/axidraw_api");   // where to do it - should be full path
  String returnedValues;                                                                     // value to return any results

  // give us some info:
  println("Running command: " + commandToRun);
  println("Location:        " + workingDir);
  println("---------------------------------------------\n");

  // run the command!
  try {

    // complicated!  basically, we have to load the exec command within Java's Runtime
    // exec asks for 1. command to run, 2. null which essentially tells Processing to 
    // inherit the environment settings from the current setup (I am a bit confused on
    // this so it seems best to leave it), and 3. location to work (full path is best)
    Process p = Runtime.getRuntime().exec(commandToRun, null, workingDir);

    // variable to check if we've received confirmation of the command
    int ii = p.waitFor();

    // if we have an output, print to screen
    if (ii == 0) {

      // BufferedReader used to get values back from the command
      BufferedReader stdInput = new BufferedReader(new InputStreamReader(p.getInputStream()));

      // read the output from the command
      while ( (returnedValues = stdInput.readLine ()) != null) {
        println(returnedValues);
      }
    }

    // if there are any error messages but we can still get an output, they print here
    else {
      BufferedReader stdErr = new BufferedReader(new InputStreamReader(p.getErrorStream()));
      println("error:");
      // if something is returned (ie: not null) print the result
      while ( (returnedValues = stdErr.readLine ()) != null) {
        println(returnedValues);
      }
    }
  }

  // if there is an error, let us know
  catch (Exception e) {
    println("Error running command!");  
    println(e);
  }

  // when done running command, quit
  println("\n---------------------------------------------");
  println("DONE!");
}
```
Here my file structure:

<img width="435" alt="Screen Shot 2020-10-02 at 5 56 08 PM" src="https://user-images.githubusercontent.com/1054816/94973004-97228100-04d8-11eb-829b-d1b979fb926f.png">

**Comment by oskay** [comment]
I would think that you should be able to use something like 
```String commandToRun = "axicli /Users/stephanschulz/Documents/axidraw-master/cli/AxiDraw_trivial.svg";```

Does that work?

**Comment by stephanschulz** [comment]
unfortunately not.

when running this `axicli version` I get this error

```
Error running command!
java.io.IOException: Cannot run program "axicli" (in directory "/Users/stephanschulz/Documents/Processing/shell/test_axicli4/axidraw_api"): error=2, No such file or directory
```

**Comment by oskay** [comment]
For me, `which axicli` returns `/usr/local/bin/axicli`, so you might try that.

**Comment by stephanschulz** [comment]
calling `which axicli` from the terminal returns `/Library/Frameworks/Python.framework/Versions/3.8/bin/axicli`

from within processing I get just `error:` :(

and `/usr/local/bin/axicli` does not exists on my macOS 10.14.6

**Comment by stephanschulz** [comment]
`  String commandToRun = "/Library/Frameworks/Python.framework/Versions/3.8/bin/axicli "+sketchpath+"data/AxiDraw_trivial.svg -v --report_time";`

also return nothing, not even an error.

**Comment by oskay** [comment]
Does `axicli` work for you from a terminal?

**Comment by stephanschulz** [comment]
yes. 
and it also works from processing. I was able to call it to send svgs to the plotter. 
but now I would also like to get my hands on the console returns so I can do something with the plotting duration info.

**Comment by oskay** [comment]
OK, LOTS of different things going on here.

(1) You say now that `axicli` works from a terminal, and also from within processing.

You also said that `axicli version` results in `Error running command![...]`

Can you please clarify to what extent `axicli` is working for you? If the version command returns an error, that doesn't sound like it's working to me.



(2) I would never expect something of the form "python axicli.py" to work unless you are in the working directory with the axicli.py file.

If you want to use axicli from the command line, then either of the following should normally work:

`axicli file.svg`
`python -m axicli file.svg`or maybe `python3 -m axicli file.svg`

If one of those would work for you, then you wouldn't need the duplicate copy of the API bundle in your processing sketch folder.


(3)
>but now I would also like to get my hands on the console returns so I can do something with the plotting duration info.

That sounds like the second issue that you have reported ( https://github.com/evil-mad/axidraw/issues/91 ), and is not actually related to this issue, or am I missing something?

**Comment by oskay** [comment]
> axicli file.svg
> python -m axicli file.svgor maybe python3 -m axicli file.svg

OK, those totally don't work in Processing -- probably need to set some environment variables.

**Comment by stephanschulz** [comment]
first of all I want to thank you for your assistance. sorry about the rather stream of conscience posts. I am just in the middle of learning all this stuff.

In terminal everything works fine.
<img width="1190" alt="Screen Shot 2020-10-02 at 7 13 55 PM" src="https://user-images.githubusercontent.com/1054816/94976829-326d2380-04e4-11eb-822a-8b42cab08fcb.png">

In processing everything works BUT not all commands are able to be printing their returns in the console.
<img width="1636" alt="Screen Shot 2020-10-02 at 7 18 12 PM" src="https://user-images.githubusercontent.com/1054816/94976850-457ff380-04e4-11eb-9266-e7fce1b565fb.png">

**Comment by stephanschulz** [comment]
Here my code:
[test_axicli5_1.zip](https://github.com/evil-mad/axidraw/files/5320720/test_axicli5_1.zip)

**Comment by oskay** [comment]
OK, so it sounds like _this_ issue is resolved, to the extent possible without setting the environment variables.

**Comment by stephanschulz** [comment]
not really. because my problem was and still is that I am not able to get the returned values when calling `... -v --report_time` from within processing.
the above image of the processing sketch show no time info being returned/ printed in the console.

**Comment by oskay** [comment]
I do not understand.

Can you please clarify:
* How is this problem related to the one reported here (" can't open file 'axicli'")
* How is this problem different from issue #91? My (perhaps naive) reading of this ("I am not able to get the returned values") is that it is identical to #91.

**Comment by stephanschulz** [comment]
Sorry, yes its the same problem as #91 
Maybe not the best GitHub abd etiquette to not stick to the correct thread.

The "can't open file 'axicli" problem within processing had to do with me needing to point my sketch to the axidraw_api folder which is located in the sketch's folder.

My processing sketch folder contains an axidraw_api folder.
https://user-images.githubusercontent.com/1054816/94973004-97228100-04d8-11eb-829b-d1b979fb926f.png


Here is the line of code that is now pointing to the correct api path: https://gist.github.com/stephanschulz/8b14e69bf17fd120d9cf51a6eab348f2#file-test_axicli-L46



After solving this I was able to run axicli commands from with in processing; like asking for version, like plotting svgs etc. But I noticed that I was not able to get a return info when wanting to know the duration of the plotting. 
So, as a simple test I used processing to ask for the axicli version. This as in the above processing screen shot, was returning the correct info.

**Comment by oskay** [comment]
This all sounds like it's still #91, not about this issue, yes?

**Comment by stephanschulz** [comment]
you are correct.
I closed issue #90 because I solved the problem.
I am now noticing that you responsed to me after I closed the issue. This must have confused me in to think I was on #91 page.
So, all posts after closing #90 should have been posted in issue #91.

**Comment by oskay** [comment]
Every question that I asked _was_ about this issue, #90.

**Comment by stephanschulz** [comment]
I thank you for your efforts and attention. Sorry about the confusions. 
I will let it sit for a few days and return to it then. 
For me #90 and #91 are all about getting the same amount of access to axicli from within processing as we have from within the terminal. 
- One problem was to have to include the axidraw_api folder in the sketch's folder.
- and other was to `File workingDir = new File(sketchpath+"/axidraw_api");` point to this folder
This has been solved to the point that I can access/use axicli.
- the only problem remaining is not getting the return values in the processing console. 

In the future I will post in #91 once I have gathered more details.
